web/auth/auth.go

```diff
@@ -177,145 +144,148 @@ func sessionData(session *sessions.Session) string {
 	return fmt.Sprintf("Session: ID=%s, IsNew=%t, %s", session.ID, session.IsNew, out)
 }
 
-// OAuth2Login tries to authenticate against an oauth2 provider.
-func OAuth2Login(logger *zap.SugaredLogger, db database.Database) echo.HandlerFunc {
-	return func(c echo.Context) error {
-		w := c.Response()
-		r := c.Request()
+// OAuth2Login redirects user to the provider's sign in page or, if user is already signed in with provider,
+// authenticates the user in the background.
+func OAuth2Login(logger *zap.SugaredLogger, authConfig *AuthConfig, secret string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		fmt.Println("AUTH2LOGIN started with: ", r.Method) // tmp
+		if r.Method != "GET" {
+			authenticationError(logger, w, fmt.Sprintf("illegal request method: %s", r.Method))
+			return
+		}
 
-		provider, err := gothic.GetProviderName(r)
+		// Start or refresh the session.
+		// Issue: server generates a new random key to encode sessions on each restart.
+		// A session from before the restart will be detected, but cannot be decoded with
+		// the new key, resulting in an error. Instead of returning, we attempt to refresh the session first.
+		s, err := sessionStore.Get(r, SessionKey)
 		if err != nil {
-			logger.Error(err.Error())
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
-		}
-		var teacher int
-		if strings.HasSuffix(provider, TeacherSuffix) {
-			teacher = 1
+			if err := sessionStore.Save(r, w, s); err != nil {
+				authenticationError(logger, w, fmt.Sprintf("failed to create new session (%s): %s", SessionKey, err))
+				return
+			}
 		}
-		logger.Debugf("Provider: %v ; Teacher: %v", provider, teacher)
 
-		qv := r.URL.Query()
-		logger.Debugf("qv: %v", qv)
-		redirect := extractRedirectURL(r, Redirect)
-		logger.Debugf("redirect: %v", redirect)
-		// TODO: Add a random string to protect against CSRF.
-		qv.Set(State, strconv.Itoa(teacher)+redirect)
-		logger.Debugf("State: %v", strconv.Itoa(teacher)+redirect)
-		r.URL.RawQuery = qv.Encode()
-		logger.Debugf("RawQuery: %v", r.URL.RawQuery)
+		// Get provider name.
+		provider := GetProviderName(r.URL.Path, 2)
+		if provider == "" {
+			authenticationError(logger, w, "incorrect request URL")
+			return
+		}
+		providerConfig, ok := authConfig.providers[provider]
+		if !ok {
+			authenticationError(logger, w, "provider is not enabled")
+			return
+		}
 
-		url, err := gothic.GetAuthURL(w, r)
-		if err != nil {
-			logger.Error(err.Error())
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+		// Check teacher suffix, update scopes if teacher.
```

Comment:
Similar for this code; seems reasonable that you could make things be a method:
```go
func (ac *AuthConfig) setScopes(urlPath string) {}
```
(but maybe not necessary to fix this if it will be deleted later anyway.

Reply:
Yes, we won't need to adjust scopes with a GitHub App.